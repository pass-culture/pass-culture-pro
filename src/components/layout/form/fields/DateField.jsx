/* eslint
  react/jsx-one-expression-per-line: 0 */
import classnames from 'classnames'
import moment from 'moment'
import React from 'react'
import PropTypes from 'prop-types'
import DatePicker from 'react-datepicker'
import { Field } from 'react-final-form'
import { composeValidators } from 'react-final-form-utils'

import Icon from 'components/layout/Icon'
import validateRequiredField from 'components/layout/form/utils/validateRequiredField'

const renderReadOnlyInput = readOnlyValue => (
  <input className="field-input field-date" readOnly value={readOnlyValue} />
)

const renderDateInput = DatePickerProps => (
  <div className="flex-columns items-center field-input field-date">
    <DatePicker {...DatePickerProps} />
    <div className="flex-auto" />
    <Icon alt="Horaires" svg="ico-calendar" />
  </div>
)

export const DateField = ({
  autoComplete,
  className,
  clearable,
  dateFormat,
  datePickerClassName,
  disabled,
  id,
  label,
  locale,
  name,
  placeholder,
  readOnly,
  renderInner,
  renderValue,
  required,
  type,
  validate,
  // see github.com/Hacker0x01/react-datepicker/blob/master/docs/datepicker.md
  ...DatePickerProps
}) => {
  const requiredIsAFunction = required && typeof required === 'function'
  const defaultRequiredValidate =
    (required && validateRequiredField) || undefined
  const requiredValidate = requiredIsAFunction
    ? required
    : defaultRequiredValidate

  const inputName = id || name
  const isClearable = !readOnly && clearable

  return (
    <Field
      name={name}
      validate={composeValidators(validate, requiredValidate)}
      render={({ input, meta }) => {
        let readOnlyValue
        let selected
        if (input.value) {
          selected = moment(input.value)
          readOnlyValue = selected.format(dateFormat)
        }
        return (
          <div
            className={classnames('field date-field', className, {
              'is-read-only': readOnly,
            })}
            id={id}>
            <label
              htmlFor={name}
              className={classnames('field-label', { empty: !label })}>
              {label && (
                <span>
                  <span>{label}</span>
                  {required && !readOnly && (
                    <span className="field-asterisk">*</span>
                  )}
                </span>
              )}
            </label>
            <div className="field-control">
              <div className="field-value flex-columns items-center">
                <div className="field-inner flex-columns items-center">
                  {readOnly && renderReadOnlyInput(readOnlyValue)}
                  {!readOnly &&
                    renderDateInput({
                      className: 'date',
                      dateFormat,
                      id: inputName,
                      isClearable,
                      locale,
                      placeholderText: placeholder,
                      shouldCloseOnSelect: true,
                      selected,
                      ...DatePickerProps,
                      ...input,
                      onChange: date => {
                        const changedValue = date ? date.toISOString() : null
                        input.onChange(changedValue)
                      },
                      value: readOnlyValue,
                    })}
                </div>
                {renderValue()}
              </div>
            </div>
          </div>
        )
      }}
    />
  )
}

DateField.defaultProps = {
  autoComplete: false,
  className: '',
  clearable: false,
  dateFormat: 'DD/MM/YYYY',
  datePickerClassName: 'date',
  disabled: false,
  id: null,
  label: '',
  locale: 'fr',
  placeholder: 'Please enter a value',
  readOnly: false,
  renderValue: () => null,
  required: false,
  validate: null,
}

DateField.propTypes = {
  autoComplete: PropTypes.bool,
  className: PropTypes.string,
  clearable: PropTypes.bool,
  dateFormat: PropTypes.string,
  datePickerClassName: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  locale: PropTypes.string,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  validate: PropTypes.func,
}

export default DateField
