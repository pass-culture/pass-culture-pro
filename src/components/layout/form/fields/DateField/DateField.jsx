import classnames from 'classnames'
import moment from 'moment-timezone'
import PropTypes from 'prop-types'
import React, { PureComponent } from 'react'
import DatePicker from 'react-datepicker'
import { Field } from 'react-final-form'
import { composeValidators } from 'react-final-form-utils'

import { InputWithCalendar } from './InputWithCalendar'
import getRequiredValidate from '../../utils/getRequiredValidate'

class DateField extends PureComponent {
  constructor(props) {
    super(props)
    this.state = { hasPressedDelete: false }
  }

  applyTimezoneToDatetime = (inputValue, timezone) => {
    let selectedDatetime = moment(inputValue)

    if (timezone) {
      selectedDatetime = selectedDatetime.tz(timezone)
    }

    return selectedDatetime
  }

  formatSelectedDatetime = (inputValue, dateFormat) => moment(inputValue).format(dateFormat)

  onDateChange = input => date => {
    const changedValue = date ? date.toISOString() : null
    input.onChange(changedValue)
    this.setState({ hasPressedDelete: false })
  }

  onKeyDown = input => event => {
    const isDeleteKey = event.keyCode === 8
    if (!isDeleteKey) return

    input.onChange(null)
    this.setState({ hasPressedDelete: true })
  }

  renderField = ({ input }) => {
    const {
      className,
      clearable,
      dateFormat,
      id,
      label,
      locale,
      name,
      placeholder,
      readOnly,
      renderValue,
      required,
      timezone,
      // see github.com/Hacker0x01/react-datepicker/blob/master/docs/datepicker.md
      ...DatePickerProps
    } = this.props
    const inputName = id || name
    const isClearable = !readOnly && clearable
    const { hasPressedDelete } = this.state
    let formattedSelectedDatetime = null
    let selectedDatetime = null

    const shouldFormatDatetime = !hasPressedDelete && input.value
    if (shouldFormatDatetime) {
      selectedDatetime = this.applyTimezoneToDatetime(input.value, timezone)
      formattedSelectedDatetime = this.formatSelectedDatetime(selectedDatetime, dateFormat)
    }

    const dateInputProps = {
      className: 'date',
      dateFormat,
      id: inputName,
      isClearable,
      locale,
      placeholderText: placeholder,
      selected: selectedDatetime,
      shouldCloseOnSelect: true,
      ...DatePickerProps,
      ...input,
      onChange: this.onDateChange(input),
      onKeyDown: this.onKeyDown(input),
      value: formattedSelectedDatetime,
      customInput: <InputWithCalendar />,
    }

    return (
      <div
        className={classnames('field date-field', className, {
          'is-read-only': readOnly,
        })}
        id={id}
      >
        <label
          className={classnames('field-label', { empty: !label })}
          htmlFor={name}
        >
          {label && (
            <span>
              <span>
                {label}
              </span>
              {required && !readOnly && (
                <span className="field-asterisk">
                  {'*'}
                </span>
              )}
            </span>
          )}
        </label>
        <div className="field-control">
          <div className="field-value flex-columns items-center">
            <div className="field-inner flex-columns items-center">
              {readOnly ? (
                <input
                  className="field-input field-date"
                  name={name}
                  readOnly
                  value={formattedSelectedDatetime}
                />
              ) : (
                <DatePicker {...dateInputProps} />
              )}
            </div>
            {renderValue()}
          </div>
        </div>
      </div>
    )
  }

  render() {
    const { name, required, validate } = this.props

    return (
      <Field
        name={name}
        render={this.renderField}
        validate={composeValidators(validate, getRequiredValidate(required))}
      />
    )
  }
}

DateField.defaultProps = {
  className: '',
  clearable: false,
  dateFormat: 'DD/MM/YYYY',
  datePickerClassName: 'date',
  disabled: false,
  id: null,
  label: '',
  locale: 'fr',
  placeholder: 'Please enter a value',
  readOnly: false,
  renderValue: function () {
    return null
  },
  required: false,
  timezone: null,
  validate: null,
}

DateField.propTypes = {
  className: PropTypes.string,
  clearable: PropTypes.bool,
  dateFormat: PropTypes.string,
  datePickerClassName: PropTypes.string,
  disabled: PropTypes.bool,
  id: PropTypes.string,
  label: PropTypes.string,
  locale: PropTypes.string,
  name: PropTypes.string.isRequired,
  placeholder: PropTypes.string,
  readOnly: PropTypes.bool,
  renderValue: PropTypes.func,
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.func]),
  timezone: PropTypes.string,
  validate: PropTypes.func,
}

export default DateField
