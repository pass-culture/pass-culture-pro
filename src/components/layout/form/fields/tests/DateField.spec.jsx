/* eslint-disable no-use-before-define */
import { mount, shallow } from 'enzyme'
import moment from 'moment'
import React from 'react'
import { Form } from 'react-final-form'

import { DateField } from '../DateField'

describe('src | components | layout | form | DateField', () => {
  it('should match snapchot', () => {
    // given
    const initialValues = {
      myDate: '2019-04-27T20:00:00Z',
    }

    // when
    const wrapper = shallow(
      <Form
        initialValues={initialValues}
        onSubmit={() => ({})}
        render={({ handleSubmit }) => (
          <form>
            <DateField name="myDate" />
            <button onClick={handleSubmit} type="submit">
              Submit
            </button>
          </form>
        )}
      />
    )

    // then
    expect(wrapper).toBeDefined()
    expect(wrapper).toMatchSnapshot()
  })

  it('should submit a form with a date', done => {
    // given
    const initialValues = {
      myDate: '2019-04-27T20:00:00Z',
    }
    const wrapper = mount(
      <Form
        initialValues={initialValues}
        onSubmit={onSubmit}
        render={({ handleSubmit }) => (
          <form>
            <DateField name="myDate" />
            <button onClick={handleSubmit} type="submit">
              Submit
            </button>
          </form>
        )}
      />
    )

    // when
    wrapper
      .find(DateField)
      .find('input[name="myDate"]')
      .simulate('click')
    wrapper
      .find('DatePicker')
      .props()
      .onChange(moment('2019-04-28T20:00:00Z'))
    wrapper.find('button[type="submit"]').simulate('click')

    // then
    function onSubmit(formValues) {
      expect(formValues.myDate).toEqual('2019-04-28T20:00:00.000Z')
      done()
    }
  })

  it('should display the date taking in account the timezone', () => {
    // given
    const initialValues = {
      myDate: '2019-04-27T02:00:00Z',
    }

    // when
    const wrapper = mount(
      <Form
        initialValues={initialValues}
        onSubmit={() => null}
        render={({ handleSubmit }) => (
          <form>
            <DateField name="myDate" readOnly timezone="America/Cayenne" />
            <button onClick={handleSubmit} type="submit">
              Submit
            </button>
          </form>
        )}
      />
    )

    // when
    expect(
      wrapper
        .find(DateField)
        .find('input[name="myDate"]')
        .props().value
    ).toEqual('26/04/2019')
  })
})
