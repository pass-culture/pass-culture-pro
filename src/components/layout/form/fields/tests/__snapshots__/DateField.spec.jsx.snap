// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src | components | layout | form | DateField should match snapchot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ReactFinalForm
    initialValues={
      Object {
        "myDate": "2019-04-27T20:00:00Z",
      }
    }
    onSubmit={[Function]}
    render={[Function]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": <form>
        <DateField
          autoComplete={false}
          className=""
          clearable={false}
          dateFormat="DD/MM/YYYY"
          datePickerClassName="date"
          disabled={false}
          id={null}
          label=""
          locale="fr"
          name="myDate"
          placeholder="Please enter a value"
          readOnly={false}
          renderValue={[Function]}
          required={false}
          timezone={null}
          validate={null}
        />
        <button
          onClick={[Function]}
          type="submit"
        >
          Submit
        </button>
      </form>,
      "value": Object {
        "batch": [Function],
        "blur": [Function],
        "change": [Function],
        "destroyOnUnregister": false,
        "focus": [Function],
        "getFieldState": [Function],
        "getRegisteredFields": [Function],
        "getState": [Function],
        "initialize": [Function],
        "isValidationPaused": [Function],
        "mutators": Object {},
        "pauseValidation": [Function],
        "registerField": [Function],
        "reset": [Function],
        "resetFieldState": [Function],
        "resumeValidation": [Function],
        "setConfig": [Function],
        "submit": [Function],
        "subscribe": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <DateField
            autoComplete={false}
            className=""
            clearable={false}
            dateFormat="DD/MM/YYYY"
            datePickerClassName="date"
            disabled={false}
            id={null}
            label=""
            locale="fr"
            name="myDate"
            placeholder="Please enter a value"
            readOnly={false}
            renderValue={[Function]}
            required={false}
            timezone={null}
            validate={null}
          />,
          <button
            onClick={[Function]}
            type="submit"
          >
            Submit
          </button>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "autoComplete": false,
            "className": "",
            "clearable": false,
            "dateFormat": "DD/MM/YYYY",
            "datePickerClassName": "date",
            "disabled": false,
            "id": null,
            "label": "",
            "locale": "fr",
            "name": "myDate",
            "placeholder": "Please enter a value",
            "readOnly": false,
            "renderValue": [Function],
            "required": false,
            "timezone": null,
            "validate": null,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Submit",
            "onClick": [Function],
            "type": "submit",
          },
          "ref": null,
          "rendered": "Submit",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": Object {
      "$$typeof": Symbol(react.provider),
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": Object {
          "$$typeof": Symbol(react.context),
          "_calculateChangedBits": null,
          "_context": [Circular],
        },
        "Provider": [Circular],
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": undefined,
        "_currentValue2": undefined,
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <form>
          <DateField
            autoComplete={false}
            className=""
            clearable={false}
            dateFormat="DD/MM/YYYY"
            datePickerClassName="date"
            disabled={false}
            id={null}
            label=""
            locale="fr"
            name="myDate"
            placeholder="Please enter a value"
            readOnly={false}
            renderValue={[Function]}
            required={false}
            timezone={null}
            validate={null}
          />
          <button
            onClick={[Function]}
            type="submit"
          >
            Submit
          </button>
        </form>,
        "value": Object {
          "batch": [Function],
          "blur": [Function],
          "change": [Function],
          "destroyOnUnregister": false,
          "focus": [Function],
          "getFieldState": [Function],
          "getRegisteredFields": [Function],
          "getState": [Function],
          "initialize": [Function],
          "isValidationPaused": [Function],
          "mutators": Object {},
          "pauseValidation": [Function],
          "registerField": [Function],
          "reset": [Function],
          "resetFieldState": [Function],
          "resumeValidation": [Function],
          "setConfig": [Function],
          "submit": [Function],
          "subscribe": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <DateField
              autoComplete={false}
              className=""
              clearable={false}
              dateFormat="DD/MM/YYYY"
              datePickerClassName="date"
              disabled={false}
              id={null}
              label=""
              locale="fr"
              name="myDate"
              placeholder="Please enter a value"
              readOnly={false}
              renderValue={[Function]}
              required={false}
              timezone={null}
              validate={null}
            />,
            <button
              onClick={[Function]}
              type="submit"
            >
              Submit
            </button>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "autoComplete": false,
              "className": "",
              "clearable": false,
              "dateFormat": "DD/MM/YYYY",
              "datePickerClassName": "date",
              "disabled": false,
              "id": null,
              "label": "",
              "locale": "fr",
              "name": "myDate",
              "placeholder": "Please enter a value",
              "readOnly": false,
              "renderValue": [Function],
              "required": false,
              "timezone": null,
              "validate": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Submit",
              "onClick": [Function],
              "type": "submit",
            },
            "ref": null,
            "rendered": "Submit",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": Object {
        "$$typeof": Symbol(react.provider),
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": Object {
            "$$typeof": Symbol(react.context),
            "_calculateChangedBits": null,
            "_context": [Circular],
          },
          "Provider": [Circular],
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": undefined,
          "_currentValue2": undefined,
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
